src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs

```diff
@@ -73,33 +73,29 @@ private PingReply SendWithPingUtility(IPAddress address, byte[] buffer, int time
 
         private async Task<PingReply> SendWithPingUtilityAsync(IPAddress address, byte[] buffer, int timeout, PingOptions? options)
         {
-            using (Process p = GetPingProcess(address, buffer, timeout, options))
-            {
-                var processCompletion = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
-                p.EnableRaisingEvents = true;
-                p.Exited += (s, e) => processCompletion.SetResult();
-                p.Start();
+            CancellationToken timeoutOrCancellationToken = _timeoutOrCancellationSource!.Token;
 
-                try
-                {
-                    await processCompletion.Task.WaitAsync(TimeSpan.FromMilliseconds(timeout)).ConfigureAwait(false);
-                }
-                catch (TimeoutException)
-                {
-                    p.Kill();
-                    return CreatePingReply(IPStatus.TimedOut);
-                }
+            using Process pingProcess = GetPingProcess(address, buffer, timeout, options);
+            pingProcess.Start();
 
-                try
+            try
+            {
+                await pingProcess.WaitForExitAsync(timeoutOrCancellationToken).ConfigureAwait(false);
+
+                string stdout = await pingProcess.StandardOutput.ReadToEndAsync(timeoutOrCancellationToken).ConfigureAwait(false);
```

Comment:
I figured it didn't hurt to also allow timing out / canceling while reading the output, but  calling it out in case it isn't desirable.